Support for Open LLMs

zigabrencic opened the pull request:
This PR resolves the [issue 943](https://github.com/gpt-engineer-org/gpt-engineer/issues/943) by adding an explanation on how to use `openLLMs`.

Overview of changes:

- minor change in `gpt_engineer/applications/cli/main.py` to work with open models.
- Rewritten `docs/open_models.md`
- Added `docs/examples` to help users test if their `LLM` setup works.

Feel free to let me know if anything is unclear or has to be modified.

Including the relevant reviewers: @ErikBjare @captivus @ATheorell @viborc

zigabrencic commented:
I managed to resolve the `tox/pytest` issues by re-naming the example scripts to `openai_api_interface.py` and `langchain_interface.py`

`pytest` thought that they were "tests" 😉

TheoMcCabe commented on docs/open_models.md:
````diff
@@ -1,24 +1,118 @@
+
+```bash
+export OPENAI_API_BASE="http://localhost:8000/v1"
+export OPENAI_API_KEY="sk-xxx"
````
Is this required? Can it not just null? Seems like an arbrirary magic string

zigabrencic replied:
According to this: https://llama-cpp-python.readthedocs.io/en/latest/server/#multimodal-models yes. 

Although I'm not sure why. Since we use `llama-cpp-python` for open model inference I decided to follow their documentation.

TheoMcCabe commented on gpt_engineer/applications/cli/main.py:
```diff
@@ -76,8 +76,12 @@ def load_env_if_needed():
+
     openai.api_key = os.getenv("OPENAI_API_KEY")
 
+    if openai.api_key == "sk-xxx":
```
this doesnt look clean to me. A boolean variable called isLocalModel or LocalModel or something would be better

zigabrencic replied:
I was thinking about this quite a lot and then decided that the above is cleanest. And requires the least amount of code changes in the code base. At the price of code readability  🤷‍♂️

If we introduce `LocalModel` variable or sth similar then we need to have a list of what clarifies as a local model. There are many options for open LLMs. Model list isn't as trivial as for the OpenAI case.

In principle we could require the user to set another env variable `local_model` and then decide on that.

Not sure. I'm open to suggestions here.

viborc replied:
@TheoMcCabe, do you have any follow-up comments on @zigabrencic's explanation?

viborc commented:
@ErikBjare @captivus @ATheorell do you guys want to weigh-in on this PR, too?

ErikBjare reviewed (commented):
I haven't looked super closely, but looks good!

One thing that struck me when reading [Aider's blog](https://aider.chat/2024/03/08/claude-3.html) is that openrouter.ai has a OpenAI-compatible API which let's users dispatch requests to a wide range of models: from OpenAI and Anthropic models, to custom/open models via Together.ai and other providers.

This might be the fastest way to get started with open models, and doesn't require any special hardware. Worth mentioning in the docs imo!

zigabrencic commented:
As discussed in today's meeting I switched to `export LOCAL_MODEL=true`. 

And removed the rest except for this line: `elif os.getenv("LOCAL_MODEL"):`  which is needed to properly compute local API costs.
